File: sirius-core/src/main/java/org/starcoin/sirius/core/WithdrawalStatus.java
package org.starcoin.sirius.core;

import org.starcoin.core.BlockAddress;
import org.starcoin.core.ProtoEnum;
import org.starcoin.core.ProtobufCodec;
import org.starcoin.proto.Starcoin.ProtoWithdrawalStatus;
import org.starcoin.proto.Starcoin.ProtoWithdrawalStatusType;

import java.util.Objects;
import java.util.logging.Logger;

/**
 * Created by dqm on 2018/10/4.
 */
public class WithdrawalStatus implements ProtobufCodec<ProtoWithdrawalStatus> {

    private Logger logger = Logger.getLogger(WithdrawalStatus.class.getName());

    public WithdrawalStatus() {
    }

    public enum WithdrawalStatusType implements ProtoEnum<ProtoWithdrawalStatusType> {
        INIT(ProtoWithdrawalStatusType.WITHDRAWAL_STATUS_INIT),
        CANCEL(ProtoWithdrawalStatusType.WITHDRAWAL_STATUS_CANCEL),
        PASSED(ProtoWithdrawalStatusType.WITHDRAWAL_STATUS_PASSED),
        CONFIRMED(ProtoWithdrawalStatusType.WITHDRAWAL_STATUS_CONFIRMED);

        private ProtoWithdrawalStatusType protoType;

        WithdrawalStatusType(ProtoWithdrawalStatusType protoType) {
            this.protoType = protoType;
        }

        @Override
        public ProtoWithdrawalStatusType toProto() {
            return protoType;
        }

        @Override
        public int getNumber() {
            return protoType.getNumber();
        }

        public static WithdrawalStatusType valueOf(int type) {
            for (WithdrawalStatusType statusType : WithdrawalStatusType.values()) {
                if (statusType.getNumber() == type) {
                    return statusType;
                }
            }
            throw new IllegalArgumentException("Unsupported status type:" + type);
        }
    }

    private Withdrawal withdrawal;
    private WithdrawalStatusType type;

    public WithdrawalStatus(Withdrawal withdrawal) {
        this.withdrawal = withdrawal;
        this.type = WithdrawalStatusType.INIT;
    }

    public WithdrawalStatus(ProtoWithdrawalStatus proto) {
        this.unmarshalProto(proto);
    }

    public boolean isInit() {
        return (this.type == WithdrawalStatusType.INIT);
    }

    public boolean cancel() {
        if (isInit()) {
            this.type = WithdrawalStatusType.CANCEL;
            logger.warning("cancel succ");
            return true;
        }

        logger.warning("cancel fail : " + ((this.type == null) ? "null" : this.type.getNumber()));
        return false;
    }

    public boolean pass() {
        if (isInit()) {
            this.type = WithdrawalStatusType.PASSED;
            logger.warning("pass succ");
            return true;
        }

        logger.warning("pass fail : " + ((this.type == null) ? "null" : this.type.getNumber()));
        return false;
    }

    public boolean finish() {
        return (this.type == WithdrawalStatusType.CONFIRMED
                || (this.type == WithdrawalStatusType.CANCEL));
    }

    public boolean isPassed() {
        return (this.type == WithdrawalStatusType.PASSED);
    }

    public boolean confirm() {
        if (this.type == WithdrawalStatusType.PASSED) {
            this.type = WithdrawalStatusType.CONFIRMED;
            logger.warning("confirm succ");
            return true;
        }

        logger.warning("confirm fail : " + ((this.type == null) ? "null" : this.type.getNumber()));
        return false;
    }

    public int getEon() {
        return this.withdrawal.getPath().getEon();
    }

    public long getWithdrawalAmount() {
        return this.withdrawal.getAmount();
    }

    public BlockAddress getAddress() {
        return this.withdrawal.getAddress();
    }

    @Override
    public ProtoWithdrawalStatus marshalProto() {
        return ProtoWithdrawalStatus.newBuilder()
                .setWithdrawal(this.withdrawal.marshalProto())
                .setType(type.toProto())
                .build();
    }

    @Override
    public void unmarshalProto(ProtoWithdrawalStatus proto) {
        this.type = WithdrawalStatusType.valueOf(proto.getTypeValue());
        this.withdrawal = proto.hasWithdrawal() ? new Withdrawal(proto.getWithdrawal()) : null;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof WithdrawalStatus)) {
            return false;
        }
        WithdrawalStatus that = (WithdrawalStatus) o;
        return Objects.equals(withdrawal, that.withdrawal) && type == that.type;
    }

    @Override
    public int hashCode() {
        return Objects.hash(withdrawal, type);
    }

    public int getStatus() {
        return this.type.getNumber();
    }
}

File: sirius-core/src/main/kotlin/org/starcoin/sirius/core/WithdrawalStatus.kt
package org.starcoin.sirius.core

import kotlinx.serialization.SerialId
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import org.starcoin.proto.Starcoin
import org.starcoin.sirius.serialization.ProtobufSchema
import java.math.BigInteger

@ProtobufSchema(Starcoin.WithdrawalStatus::class)
@Serializable
data class WithdrawalStatus(
    @SerialId(1)
    var type: WithdrawalStatusType = WithdrawalStatusType.INIT,
    @SerialId(2)
    val withdrawal: Withdrawal = Withdrawal.DUMMY_WITHDRAWAL
) : SiriusObject() {

    @Transient
    val withdrawalAmount: BigInteger
        get() = this.withdrawal.amount

    @Transient
    val status: Int
        get() = this.type.number

    @Transient
    val isInit: Boolean
        get() = this.type == WithdrawalStatusType.INIT

    @Transient
    val eon: Int
        get() = this.withdrawal.proof.path.eon

    fun pass() {
        if (isInit) {
            this.type = WithdrawalStatusType.PASSED
        }
    }

    fun cancel() {
        if (isInit) {
            this.type = WithdrawalStatusType.CANCEL
        }
    }

    fun clientConfirm() {
        if (this.type == WithdrawalStatusType.PASSED || isInit || this.type == WithdrawalStatusType.CONFIRMED) {
            this.type = WithdrawalStatusType.CLIENTCONFIRMED
        }
    }

    companion object :
        SiriusObjectCompanion<WithdrawalStatus, Starcoin.WithdrawalStatus>(WithdrawalStatus::class) {

        var DUMMY_WITHDRAWAL_STATUS = WithdrawalStatus()

        override fun mock(): WithdrawalStatus {
            return WithdrawalStatus(WithdrawalStatusType.INIT, Withdrawal.mock())
        }
    }
}

